tools/codegen/gen_functionalization_type.py

```diff
@@ -40,15 +39,26 @@ def is_tensor_like(a: Union[Argument, TensorOptionsArguments, SelfArgument]) ->
 # unwraps all tensor-like arguments, returning:
 # (1) a string containing all of the logic that does the unwrapping
 # (2) a context, to be used by translate(), with all of the relevant bindings.
-def unwrap_tensor_args(sig: DispatcherSignature) -> Tuple[str, List[Binding]]:
+def unwrap_tensor_args(sig: DispatcherSignature, *, is_view_op: bool) -> Tuple[str, List[Binding]]:
     context: List[Binding] = []
     unwrapped_tensor_args: List[str] = []
     for arg in sig.arguments():
         if is_tensor_like(arg.argument):
             # for tensor inputs, we want to unwrap them before passing them into the redispatch calls.
             unwrapped_name = f'{arg.name}_'
-            unwrapped_tensor_args.append(
-                f'auto {unwrapped_name} = at::functionalization::impl::from_functional_tensor({arg.name});')
+            # For most ops, the functionalization needs to sync any pending updates on the input tensors
+            # before calling the operator, since otherwise the operator will act on stale data.
+            # For view ops though, we can continue to defer syncing until the tensor is used by
+            # a non-view operator.
+            maybe_sync_input = '' if is_view_op else f'at::functionalization::impl::sync({arg.name});'
+            unwrapped_tensor_args.append(f"""
+      {arg.nctype.remove_const_ref().cpp_type()} {unwrapped_name};
+      if (at::functionalization::impl::isFunctionalTensor({arg.name})) {{
+        {maybe_sync_input}
+        {unwrapped_name} = at::functionalization::impl::from_functional_tensor({arg.name});
+      }} else {{
+        {unwrapped_name} = {arg.name};
+      }}""")
```

Comment:
@ezyang @bdhirsh 
I think I'm having problems related to this change on #73350.

1. Whenever the argument is a `TensorList`, `{arg.nctype.remove_const_ref().cpp_type()}` turns out to be a `TensorList` (not an owning type), too. This will likely lead to segmentation faults.
2. It only compiles because [`from_functional_tensor`](https://github.com/pytorch/pytorch/blob/6642e88ad23bfcf1349f2bce06f3490e49f195bf/aten/src/ATen/FunctionalTensorWrapper.cpp#L298-L304) also returns a `TensorList`. That's a bug, though, since the returned `TensorList` is, in fact, a local `std::vector`.

Reply:
Oof, thanks for catching that (and sorry for making you find that bug). Your right, I think that the output of `from_functional_tensor` should always be an owning type (so a vector, in the case where the input was a `TensorList`).

I'm in the middle of making some changes to functionalization, so I can add that fix too and post here when it's ready.

Reply:
(fix here: https://github.com/pytorch/pytorch/pull/76125)